main_test.go

```diff
@@ -963,3 +966,168 @@ func Test_handleClusterNameResponse(t *testing.T) {
 		t.Errorf("the new key %s does not match what it should be %s", string(resp.Kvs[0].Value), "bananas")
 	}
 }
+
+const internalMetricsReportingConfig = `
+ {
+   "LogFormat": "logfmt",
+   "LogDir": "-",
+   "NumProcs":4,
+   "DebugFlag": "debugme",
+   "ListenFrom":[
+     {
+         "Type":"signalfx",
+         "ListenAddr": "127.0.0.1:0"
+     }
+   ],
+   "LocalDebugServer": "127.0.0.1:0",
+   "ForwardTo":[
+   {
+     "type": "signalfx-json",
+     "DefaultAuthToken": "AAA",
+     "url": "http://localhost:<<PORT>>/v2/datapoint",
+     "eventURL": ""
+   }
+   ],
+    "MaxGracefulWaitTime":     "10ms",
+    "GracefulCheckInterval":   "10ms",
+    "MinimalGracefulWaitTime": "10ms",
+    "SilentGracefulTime": "50ms",
+    "InternalMetricsReportingDelay": "1s"
+ }
+`
+const statsDelayConfig = `
+ {
+   "LogFormat": "logfmt",
+   "LogDir": "-",
+   "NumProcs":4,
+   "DebugFlag": "debugme",
+   "ListenFrom":[
+     {
+         "Type":"signalfx",
+         "ListenAddr": "127.0.0.1:0"
+     }
+   ],
+   "LocalDebugServer": "127.0.0.1:0",
+   "ForwardTo":[
+   {
+     "type": "signalfx-json",
+     "DefaultAuthToken": "AAA",
+     "url": "http://localhost:<<PORT>>/v2/datapoint",
+     "eventURL": ""
+   }
+   ],
+    
+    "StatsDelay": "1s"
+ }
+`
+
+func TestPrefixAddition(t *testing.T) {
+	prefixedMetrics := map[string]bool{
+		"gateway.total_process_errors": true,
+		"gateway.total_datapoints":     true,
+		"gateway.total_events":         true,
+		"gateway.total_spans":          true,
+		"gateway.total_process_calls":  true,
+		"gateway.dropped_points":       true,
+		"gateway.dropped_events":       true,
+		"gateway.dropped_spans":        true,
+		"gateway.process_time_ns":      true,
+		"gateway.calls_in_flight":      true,
+	}
+
+	Convey("should not add prefixes to metrics that are not internal to the gateway", t, func() {
+		cancelfunc, sfxGateway, sl, sendTo := setUpSfxGateway(statsDelayConfig)
+		tk := timekeepertest.NewStubClock(time.Now())
+		sfxGateway.tk = tk
+		var end bool
+		for !end {
+			select {
+			case <-time.After(*sfxGateway.config.StatsDelayDuration * 2):
+				end = true
+			}
+		}
+		sfxGateway.signalChan <- syscall.SIGTERM
+		So(sl.Close(), ShouldBeNil)
+		if cancelfunc != nil {
+			cancelfunc()
+		}
+		dps := <-sendTo.PointsChan
+		So(len(dps), ShouldNotEqual, 0)
+		for _, dp := range dps {
+			_, ok := prefixedMetrics[dp.Metric]
```

Comment:
not something you need to fix, but just education, you're using a map[string]bool, but then ignoring the bool and just checking the existence bool... kinda weird?  it is testing what you think it is, just weird.

Reply:
golang doesn't have built in set type and I wanted O(1) lookup. so ended up using a map with bool